core/src/main/java/io/questdb/cutlass/line/tcp/DelegatingTlsChannel.java

```diff
@@ -0,0 +1,367 @@
+
+        this.wrapOutputBuffer = ByteBuffer.allocateDirect(initialCapacity);
+        this.unwrapInputBuffer = ByteBuffer.allocateDirect(initialCapacity);
+        this.unwrapOutputBuffer = ByteBuffer.allocateDirect(initialCapacity);
+        this.wrapOutputBufferPtr = Unsafe.getUnsafe().getLong(wrapOutputBuffer, ADDRESS_FIELD_OFFSET);
+        this.unwrapInputBufferPtr = Unsafe.getUnsafe().getLong(unwrapInputBuffer, ADDRESS_FIELD_OFFSET);
+        this.unwrapOutputBufferPtr = Unsafe.getUnsafe().getLong(unwrapOutputBuffer, ADDRESS_FIELD_OFFSET);
+        this.dummyBuffer = ByteBuffer.allocate(0);
+    }
+
+    private static SSLEngine createSslEngine(String trustStorePath, char[] trustStorePassword) {
+        try {
+            SSLContext sslContext;
+            // intentionally not exposed to end user as an option
+            // it's used for testing, but dangerous in prod
+            if (trustStorePath != null) {
+                sslContext = SSLContext.getInstance("SSL");
+                TrustManagerFactory tmf = TrustManagerFactory.getInstance(TrustManagerFactory.getDefaultAlgorithm());
+                KeyStore jks = KeyStore.getInstance("JKS");
+
+                InputStream trustStoreStream = null;
+                try {
+                    if (trustStorePath.startsWith("classpath:")) {
+                        String adjustedPath = trustStorePath.substring("classpath:".length());
+                        trustStoreStream = DelegatingTlsChannel.class.getResourceAsStream(adjustedPath);
+                        if (trustStoreStream == null) {
+                            throw new LineSenderException("Configured trust at classpath:" + trustStorePath + " is unavailable on a classpath");
+                        }
+                    } else {
+                        trustStoreStream = new BufferedInputStream(new FileInputStream(trustStorePath));
+                    }
+                    jks.load(trustStoreStream, trustStorePassword);
+                } finally {
+                    if (trustStoreStream != null) {
+                        trustStoreStream.close();
+                    }
+                }
+                tmf.init(jks);
+                TrustManager[] trustManagers = tmf.getTrustManagers();
+                sslContext.init(null, trustManagers, new SecureRandom());
+            } else {
+                sslContext = SSLContext.getDefault();
+            }
+            SSLEngine sslEngine = sslContext.createSSLEngine();
+            sslEngine.setUseClientMode(true);
+            return sslEngine;
+        } catch (NoSuchAlgorithmException | KeyManagementException | CertificateException | KeyStoreException |
+                 IOException e) {
+            throw new LineSenderException("error while creating openssl engine", e);
+        }
+    }
+
+    @Override
+    public void send(long ptr, int len) {
+        try {
+            handshakeLoop();
```

Comment:
It would be better to do have handshake loop once, perhaps in constructor and not call it every method. This also be faster fail for the client otherwise client will fail only when try to send something.

Reply:
that sounds like a good idea! 

we want a client to finish handshake as early as possible anyway. as I can imagine e.g. HAProxy won't like a long delay between a connection being opened and TLS handshake! thanks!